include/crab/domains/object_domain.hpp

```diff
@@ -0,0 +1,2378 @@
+    // Solution: we join the base dom and odi map
+    // E.g.
+    //  s1: Base = { V_1 = 3, V_2 = 6 }, Object (V_1, V_2) = empty
+    //  s2: Base = { V_1 = 2, V_2 = 4 }, Object (V_1, V_2) = empty
+    //  s1 U s2: Base in s1 U Base in s2
+    //           Object (V_1, V_2) = empty
+
+    // Case 3:
+    // one state: singleton object in base domain,
+    //  odi map does not contain a field domain for that object
+    // another: not singleton, odi map contains a field domain for that object
+    // Solution: project singleton object, join with non-singleton in the odi
+    // map
+    //  Let s1 be the first state, s2 is second one.
+    //  Let Vs be the fields in that abstract object.
+    //  For Base in s1,
+    //  only_singleton = Base.project(Vs)
+    //  no_singleton = Base.forget(Vs)
+    //  s1 U s2 :
+    //  no_singleton U Base in s2,
+    //      Object (Vs) in s2 U only_singleton
+    // E.g.
+    //  s1: Base = { V_1 = 3, V_2 = 6 }, Object (V_1, V_2) = empty
+    //  s2: Base = { ... }, Object (V_1, V_2) = { V_1 <= 3, V_2 <= 5 }
+    //  s1 U s2:
+    //    Base = (Base in s1).project(V_1, V_2) U Base in s2
+    //    Object (V_1, V_2) = { V_1 <= 3, V_2 <= 6 }
+
+    // Other cases (simple):
+    // one state: exists such an abstract object
+    // another: does not exist that abstract object
+    // Solution: we join the base dom and odi map
+    // Example1:
+    //  s1: Base = { V_1 = 3, V_2 = 6 }, Object (V_1, V_2) = empty
+    //  s2: Base = { ... }, Object (V_1, V_2) = empty
+    //  s1 U s2 :
+    //    Base = { ... }, Object (V_1, V_2) = empty
+
+    // Example2:
+    //  s1: Base = { ... }, Object (V_1, V_2) = { V_1 <= V_2, V_2 <= 5 }
+    //  s2: Base = { ... }, Object (V_1, V_2) = empty
+    //  s1 U s2 :
+    //    Base = Base^s1 U Base^s2,
+    //    Object (V_1, V_2) = empty
+
+    // Overall, for both cases, the only third case requires move singleton
+    // into odi map. Other cases require join operation on each subdomain as
+    // before.
+
+    // 1. join the parts of each state that can use default join operation
+    obj_info_env_t out_obj_info_env(m_obj_info_env | right.m_obj_info_env);
+    address_abstract_domain_t out_addrs_dom(m_addrs_dom | right.m_addrs_dom);
+
+    // m_flds_id_map and m_refs_base_addrs_map do not require join,
+    // they share accross all states. No need to update
+    assert(m_flds_id_map == right.m_flds_id_map);
+    assert(m_refs_base_addrs_map == right.m_refs_base_addrs_map);
+
+    // 2. join the odi map
+    base_abstract_domain_t &l_base_dom = m_base_dom;
+    base_abstract_domain_t r_base_dom = right.m_base_dom;
+    odi_map_t out_odi_map;
+    for (auto kv : m_obj_info_env) {
+      const obj_id_t &id = kv.first;
+      const object_info_t &l_obj_info = kv.second;
+      const small_range &l_num_refs = l_obj_info.refcount_val();
+      auto r_obj_info_ref = right.m_obj_info_env.find(id);
+      if (!r_obj_info_ref) {
```

Comment:
If an object is not in `right` does it represent that for that object we have top? should the object be forgotten in `this`?